server/scsynth/SC_PortAudio.cpp

```diff
@@ -262,39 +270,181 @@ int SC_PortAudioDriver::PortAudioCallback(const void* input, void* output, unsig
+    streamParams.suggestedLatency = suggestedLatency;
+    streamParams.hostApiSpecificStreamInfo = nullptr;
+    return streamParams;
+}
+
+template <typename SupportCheck> // expected signature: PaError (*)(PaStreamParameters&, double)
+PaError SC_PortAudioDriver::CheckSinglePaDevice(int* device, double sampleRate, int maxChannels, int defaultDevice,
+                                                SupportCheck isSupportedFunc, const char* deviceType) const {
+    if (*device != paNoDevice && sampleRate) {
+        // check if device can support requested SR
+        PaStreamParameters parameters = GetPaStreamParameters(*device, maxChannels, 0);
+        PaError err = isSupportedFunc(parameters, sampleRate);
+        if (err != paNoError) {
+            fprintf(stdout, "PortAudio error: %s\nRequested sample rate %f for device %s is not supported\n",
+                    Pa_GetErrorText(err), sampleRate, Pa_GetDeviceInfo(*device)->name);
+            return err;
+        }
+    }
+    // in case we still don't have a proper device, use the default device
+    if (*device == paNoDevice) {
+        *device = defaultDevice;
+        if (*device != paNoDevice)
+            fprintf(stdout, "Selecting default system %s device\n", deviceType);
+    }
+    return paNoError;
+}
+
+void SC_PortAudioDriver::SelectMatchingPaDevice(int* matchingDevice, int* knownDevice,
+                                                IOType matchingDeviceType) const {
+    if (*matchingDevice != paNoDevice || *knownDevice == paNoDevice)
+        return;
+
+    const PaHostApiInfo* apiInfo;
+    apiInfo = Pa_GetHostApiInfo(Pa_GetDeviceInfo(*knownDevice)->hostApi);
+
+    bool isInput;
+    if (matchingDeviceType == IOType::Input)
+        isInput = true;
+    else if (matchingDeviceType == IOType::Output)
+        isInput = false;
+
+    if (isInput) {
+        *matchingDevice = apiInfo->defaultInputDevice;
+    } else {
+        *matchingDevice = apiInfo->defaultOutputDevice;
+    }
+    if (*matchingDevice != paNoDevice)
+        fprintf(stdout, "Selecting default %s %s device\n", apiInfo->name, (isInput ? "input" : "output"));
+}
+
+// this function will select default PA devices if they are not defined
+// it will also try to check for some configuration problems
+// numIns, numOuts and sampleRate are only the requested values, can change later
+PaError SC_PortAudioDriver::CheckPaDevices(int* inDevice, int* outDevice, int numIns, int numOuts,
+                                           double sampleRate) const {
+    if (numIns && !numOuts) {
+        *outDevice = paNoDevice;
+        // check for requested sample rate or select the default device
+        return CheckSinglePaDevice(
+            inDevice, sampleRate, (*inDevice != paNoDevice) ? Pa_GetDeviceInfo(*inDevice)->maxInputChannels : 0,
+            Pa_GetDefaultInputDevice(),
+            [](PaStreamParameters& params, double sr) { return Pa_IsFormatSupported(&params, nullptr, sr); }, "input");
```

Comment:
this is looking a lot better! I think if you move this subexpression out before the call to CheckSinglePaDevice(), it will be a little clearer:

```cpp
auto maxChannels = (*inDevice != paNoDevice) ? Pa_GetDeviceInfo(*inDevice)->maxInputChannels : 0;
return CheckSinglePaDevice(inDevice, sampleRate, maxChannels, Pa_GetDefaultInputDevice(), "input", 
    [](PaStreamParameters& params, double sr) { return Pa_IsFormatSupported(&params, nullptr, sr); });
```

same goes for the branch below